File: python3/swapping_nodes_in_a_linked_list.py
from .model import ListNode

"""
Space   : O(1)
Time    : O(n)
"""

class Solution:
    def swapNodes(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        p = head
        n = 0

        while p:
            p = p.next
            n += 1

        s = k - 1
        e = n - k
        p1, p2 = head, head

        while s > 0 or e > 0:
            if s > 0:
                p1 = p1.next
                s -= 1
            if e > 0:
                p2 = p2.next 
                e -= 1

        p1.val, p2.val = p2.val, p1.val               
        return head

File: golang/swapping_nodes_in_a_linked_list.go
package main

import (
	"golang/model"
)

// Space	: O(n)
// Time		: O(n)

func swapNodes(head *model.ListNode, k int) *model.ListNode {
	temp := []int{}
	p := head

	for p != nil {
		temp = append(temp, p.Val)
		p = p.Next
	}

	temp[k-1], temp[len(temp)-k] = temp[len(temp)-k], temp[k-1]

	p = head
	i := 0
	for i < len(temp) {
		p.Val = temp[i]
		i++
		p = p.Next
	}

	return head
}
